bgpd/bgp_io.c

```diff
@@ -235,6 +236,7 @@ static int bgp_process_reads(struct thread *thread)
 			stream_set_endp(pkt, pktsize);
 
 			frr_with_mutex(&peer->io_mtx) {
+				frrtrace(2, frr_bgp, packet_read, peer, pkt);
```

Comment:
shouldn't this be protected by an enable check (same comment below *not* repeated)

Reply:
No, this is one of the primary features of tracepoints. They only execute if they are enabled by some external observer. When using LTTng, a tracepoint expands to e.g. this:
```
 do {      
         ;                                                                                                                                                      
         if (__builtin_expect(!!(__extension__({                           
                     __asm__ __volatile__("" : : : "memory");                                                                                                                                    
                     (*(__volatile__ __typeof__(      
                             __tracepoint_frr_libfrr___thread_call.state)      
                                *)&(__tracepoint_frr_libfrr___thread_call      
                                            .state));                   
             })),               
                              0))                                                   
                 __tracepoint_cb_frr_libfrr___thread_call(                   
                         thread->master, thread->funcname, thread->schedfrom,      
                         thread->schedfrom_line, ((void *)0), thread->u.fd,      
                         thread->u.val, thread->arg, thread->u.sands.tv_sec);      
 } while (0);   
```

When using USDT, they compile into a NOP that is overwritten at runtime when a tracer enables the probe:
```
Breakpoint 1, main () at main_usdt.c:22
22		frrtrace(2, qlyoung_usdt, calculated, x, y);
(gdb) disassemble
Dump of assembler code for function main:
   0x000055555555471a <+0>:	push   %rbp
   0x000055555555471b <+1>:	mov    %rsp,%rbp
   0x000055555555471e <+4>:	sub    $0x10,%rsp
   0x0000555555554722 <+8>:	callq  0x5555555545d0 <getchar@plt>
   0x0000555555554727 <+13>:	movl   $0x14,-0x8(%rbp)
   0x000055555555472e <+20>:	movl   $0x1e,-0x4(%rbp)
   0x0000555555554735 <+27>:	callq  0x5555555545f0 <random@plt>
   0x000055555555473a <+32>:	mov    %eax,%edx
   0x000055555555473c <+34>:	mov    -0x8(%rbp),%eax
   0x000055555555473f <+37>:	add    %edx,%eax
   0x0000555555554741 <+39>:	mov    %eax,-0x8(%rbp)
=> 0x0000555555554744 <+42>:	nop
   0x0000555555554745 <+43>:	mov    0x2008d4(%rip),%rax        # 0x555555755020 <stderr@@GLIBC_2.2.5>
   0x000055555555474c <+50>:	mov    -0x8(%rbp),%edx
   0x000055555555474f <+53>:	lea    0x9e(%rip),%rsi        # 0x5555555547f4
   0x0000555555554756 <+60>:	mov    %rax,%rdi
   0x0000555555554759 <+63>:	mov    $0x0,%eax
   0x000055555555475e <+68>:	callq  0x5555555545e0 <fprintf@plt>
   0x0000555555554763 <+73>:	mov    $0x0,%eax
   0x0000555555554768 <+78>:	leaveq 
   0x0000555555554769 <+79>:	retq   
End of assembler dump.
```

You'll notice some tracepoints, e.g. the ones in `bgp_route.c`, do have `frrtrace_enabled()` guards around them. This is only necessary when you have some additional preprocessing work you want to do to generate the data that is exported in the trace event, and you want to avoid that work when the tracepoint isn't enabled. Usually you don't need to do this because you can express that work directly in the tracepoint definition, but what you can do in there is limited by what serialization functions LTTng provides (actually there is a way to add arbitrary code to the generated serialization function but this is complicated and not really recommended).

Reply:
so what is frrtrace defined as when tracing is not present?

Reply:
https://github.com/FRRouting/frr/pull/7194/files#diff-e275b517c623e0b02fced60d8028a0c4R72